systemd zvol target

aerusso opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->
<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This is an alternative to #8975, implementing granular zfs-volume@.service units for each zvol.

### Description
<!--- Describe your changes in detail -->
Introduce a zfs-volumes.target, which Requires= zfs-volume@.service units that are dynamically generated for each zvol by the existing systemd-generator infrastructure at boot time.

Each unit polls once per second, exiting if it detects the zvol link, succeeding unconditionally after a 30 second timeout.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
It builds. The target takes 30 seconds to start if there is a missing link to a zvol, starts up within a second of the link being created. Appropriate systemd links were created after a `systemctl daemon-reload`.  The modified version of zfs-list.cache did not cause interference with mismatched versions of zfs-mount-generator.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

aerusso commented on etc/systemd/system/zfs-volume@.service.in:
```diff
@@ -0,0 +1,9 @@
+DefaultDependencies=no
+
+[Service]
+ExecStart=sh -c 'timeout=30 ; while [ $timeout -gt 0 ] ; do [ -e "/dev/zvol/$0" ] && exit 0 ; sleep 1 ; timeout=$((timeout-1)); done ; echo "zvol $0 unavailable"' %I
```
The man page (in Debian and [Ubuntu](https://manpages.ubuntu.com/manpages/cosmic/man5/systemd.service.5.html)), says that "It is thus safe to use just the executable name in case of executables located in any of the "standard" directories".  It does recommend absolute paths to "avoid ambiguity". Could you confirm that this is actually broken in Ubuntu? If there's even one example where this could be a problem, I think it's worth adjusting.

c0d3z3r0 replied:
Omfg. Ubuntu 12.04, 18.10, 19.04 allow non-absolute paths while 14.04, 16.04, 18.04 requires absolute paths (they fail with *Executable path is not absolute, ignoring: sh -c*). WTF are they doing...

Short: yes, ExecStart must be absolute here

pzakha replied:
Thanks for testing this @c0d3z3r0. Sorry I missed your comment @aerusso. This is indeed madness...

aerusso replied:
@c0d3z3r0 Strange! Thanks for testing, I've updated this to use `/bin/sh`.

aerusso commented:
Thanks for looking at this @pzakha!

> Not related to this code, but I've noticed that /usr/lib/systemd/system-generators/zfs-mount-generator introduced in #7329 is not a valid path for generators on Ubuntu, so I had to move this to /lib/systemd/system-generators/zfs-mount-generator to have this work. I'll file a separate bug for that.

That's a packaging issue as far I can tell.  The Debian source (which I've made sure stayed synchronized) has the correct settings (in `dh_auto_configure`). It might make sense to give more sensible defaults, but I think the feeling I've gotten on these kinds of issues is that it should be solved by packagers (though I can see merit in zfsonlinux doing distribution-specific builds, I've decided to draw the battle lines at native Debian packaging).

> I've tested this on the same system as for #8975, which had ~1000 zvols. ...

I was worried about this, actually. Earlier version of the zed scriptlet actually tried to only update the affected datasets. We dropped it at the time because it was getting complicated. Perhaps it can be revisited, given the demonstrated need.

> To keep the naming a bit more consistent, I can rename `zfs-volume.target` to `zfs-volumes.target`

I feel like `zfs-volumes.target` makes more sense, so I think that's for the best!

behlendorf commented:
@aerusso @pzakha I think we should strongly consider pulling this change and #8975 in to a single PR.  This would let us easily resolve any discrepancies, for example we'll need to settle on `zfs-volumes.target` or `zfs-volume.target`.  Additionally, if we want this new target to be installed with ` WantedBy=zfs.target`.  This seems reasonable to me, but in this regard the two PRs still differ.

My suggestion would be to lay this out the same way as the `zpool-import.target`, so either the `zfs-volume-wait.service` or `zfs-volume@.service` could be used by `zfs-volumes.target`.  Defaulting to `zfs-volume-wait.service` I think makes the most sense since it requires no additional configuration.  As does disabling the `zfs-volumes.target`, at least initially.  So:

```
enable zfs-volume-wait.service
disable zfs-volume@.service
disable zfs-volumes.target 
```

Does that sound like a reasonable way forward for both of these PRs?

pzakha commented:
> @aerusso @pzakha I think we should strongly consider pulling this change and #8975 in to a single PR. This would let us easily resolve any discrepancies, for example we'll need to settle on `zfs-volumes.target` or `zfs-volume.target`. Additionally, if we want this new target to be installed with ` WantedBy=zfs.target`. This seems reasonable to me, but in this regard the two PRs still differ.
> 
> My suggestion would be to lay this out the same way as the `zpool-import.target`, so either the `zfs-volume-wait.service` or `zfs-volume@.service` could be used by `zfs-volumes.target`. Defaulting to `zfs-volume-wait.service` I think makes the most sense since it requires no additional configuration. As does disabling the `zfs-volumes.target`, at least initially. So:
> 
> ```
> enable zfs-volume-wait.service
> disable zfs-volume@.service
> disable zfs-volumes.target 
> ```
> 
> Does that sound like a reasonable way forward for both of these PRs?

I was planning to do 2 changes to #8975 to make it consistent with this PR:
1. rename `zfs-volume.target` -> `zfs-volumes.target`
2. Add `WantedBy=zfs.target` to `zfs-volumes.target` (but keep the target disabled).

I think given those changes, this PR could be a natural follow-up, as bundling them together in a single review might delay integration of #8975. @aerusso what do you think?

pzakha commented:
> > Not related to this code, but I've noticed that /usr/lib/systemd/system-generators/zfs-mount-generator introduced in #7329 is not a valid path for generators on Ubuntu, so I had to move this to /lib/systemd/system-generators/zfs-mount-generator to have this work. I'll file a separate bug for that.
> 
> That's a packaging issue as far I can tell. The Debian source (which I've made sure stayed synchronized) has the correct settings (in `dh_auto_configure`). It might make sense to give more sensible defaults, but I think the feeling I've gotten on these kinds of issues is that it should be solved by packagers (though I can see merit in zfsonlinux doing distribution-specific builds, I've decided to draw the battle lines at native Debian packaging).
> 
> >

Yes, you are right. It looks like we do not explicitly set `--with-systemdgeneratordir`, which should fix this problem.

pzakha commented on etc/systemd/system-generators/zfs-mount-generator.in:
```diff
@@ -46,6 +46,8 @@ fi
 # avoid regressions, this dependency is reduced to "wants" rather than
 # "requires". **THIS MAY CHANGE**
 req_dir="${dest_norm}/local-fs.target.wants/"
+zvol_reqdir="${dest_norm}/zfs-volumes.target.requires"
```
Now that we integrated https://github.com/zfsonlinux/zfs/pull/8975, do you think we still need to add an additional `requires` on `zfs-volumes.target`?

aerusso replied:
I believe so (enabling the `zfs-volume@` units should presumably allow the user to disable `zfs-volume-wait.service`).

I feel like I may be misunderstanding something.

pzakha replied:
My original idea was to only use `zfs-volume-wait.service` as a dependency for `zfs-volumes.target`, but that's mostly because using it along with `zfs-volume@` as a dependency was redundant. Disabling `zfs-volume-wait.service` when we are using `zfs-volume@` would be another way to do this.

aerusso commented:
I've rebased on master, and implemented the changes. I have not tested it to see how it interacts with #8975, but I've disabled the template units by default. Presumably, we should add some documentation about disabling `zfs-volume-wait.service` and enabling `zfs-volume@.service` somewhere. Maybe the `zfs-mount-generator` man page?

pzakha commented:
> Presumably, we should add some documentation about disabling zfs-volume-wait.service and enabling zfs-volume@.service somewhere. Maybe the zfs-mount-generator man page?

That sounds good to me. I think we may also want to mention the trade-offs between the 2 implementations, especially regarding scalability.

I'm also wondering if the 30 seconds timeout is too aggressive. If the `zfs-volume@.service` service returns "OK" when it is not effectively ready then that could have an impact on whoever depends on that zvol link to be there.

behlendorf commented:
@aerusso @c0d3z3r0 what are your thought on moving this PR forward?

c0d3z3r0 commented on etc/systemd/system/50-zfs.preset.in:
```diff
@@ -4,6 +4,8 @@ disable zfs-import-scan.service
 enable zfs-mount.service
 enable zfs-share.service
+disable zfs-volumes.target
+disable zfs-volume@.service
```
not sure if that is needed since the service only comes into effect when some zfs-volume@volname is enabled

aerusso replied:
I will gladly enable this if this is the consensus. We seem to be moving forward very cautiously with this zfs-mount-generator project, and I don't want to be too aggressive.

behlendorf replied:
I think it would be reasonable to enable this and the zfs-mount-generator by default in the next tagged release.  Why don't we go ahead and merge this so it's currently disabled, then open a new PR to enable the generators by default.  This way it'll be enabled in the master branch by default well before anything is tagged.

c0d3z3r0 replied:
I‘m ok with the target disabled but disable the template is nonsense as it does not disable anything ;-)

c0d3z3r0 replied:
so, line 8 can be dropped but 7 must remain

c0d3z3r0 replied:
Ah, one more thing... zfs-volume@.service isn't even an installable unit, so there is not need for the disabled line because it can't be enabled anyways

aerusso replied:
Yeah, `disable zfs-volume@*.service` might parse (the documentation on these preset files is pretty sparse), but it's also unnecessary, since it would only exist if you were trying to use this.

Also, line 7 should also go. I added that disable out of an abundance of caution (not creating and enabling a feature). This patch doesn't create that `.target' anymore, so I'm not going to disable it.

c0d3z3r0 commented on etc/systemd/system/zfs-volume@.service.in:
```diff
@@ -0,0 +1,9 @@
+[Service]
+ExecStart=/bin/sh -c 'timeout=30 ; while [ $timeout -gt 0 ] ; do [ -e "/dev/zvol/$0" ] && exit 0 ; sleep 1 ; timeout=$((timeout-1)); done ; echo "zvol $0 unavailable"' %I
+Type=oneshot
+RemainAfterExit=true
```
don't mix true/false/yes/no; systemd convention is yes/no

aerusso replied:
Thanks, fixed this.

c0d3z3r0 commented on etc/systemd/system-generators/zfs-mount-generator.in:
```diff
@@ -46,6 +46,8 @@ fi
 # avoid regressions, this dependency is reduced to "wants" rather than
 # "requires". **THIS MAY CHANGE**
 req_dir="${dest_norm}/local-fs.target.wants/"
+zvol_reqdir="${dest_norm}/zfs-volumes.target.requires"
```
requires will lead to a failing boot / systemd going to recovery mode - is this really what we want? why not just fail ("wants")?

aerusso replied:
`zfs-volumes.target` will fail, but `zfs.target` only `Wants=` that target, so we're good. I'm not sure what the best policy is here---the boot should not fail, if I understand what is happening correctly.

c0d3z3r0 replied:
Ok, but needs testing

c0d3z3r0 replied:
Aah, I head some knot in my brain. This is just fine ;-)

c0d3z3r0 commented on etc/systemd/system/zfs-volume@.service.in:
```diff
@@ -0,0 +1,9 @@
+DefaultDependencies=no
+
+[Service]
+ExecStart=/bin/sh -c 'timeout=30 ; while [ $timeout -gt 0 ] ; do [ -e "/dev/zvol/$0" ] && exit 0 ; sleep 1 ; timeout=$((timeout-1)); done ; echo "zvol $0 unavailable"' %I
```
this is systemd, no need for such complicated construct :-)

~~~
BindsTo=%i.device
After=%i.device
Conflicts=shutdown.target  # maybe not needed, needs testing
~~~

aerusso replied:
Wow, that's a lot nicer!

I'm not super familiar with how the timeout on this works. If this device is slow to show up, will this give up and enter a failed state, or forever remain in a waiting state? Is this what `JobTimeoutSec=` does (if so, when does it become "queued"?) The other timeout setting I see in `man systemd.unit` seems to do with once the job has fired.  Is my understanding correct?

c0d3z3r0 replied:
Sorry for my late response, I Will have a look again later rhis day

c0d3z3r0 replied:
JobTimeoutSec is there for a) setting a timeout other than the default (infinity) for normal units and b) for overriding the default device unit timeout (which is specified by DefaultTimeoutStartSec=). In our case b) applies, which we should not modify upstream. DefaultTimeoutStartSec is 90s for Debian for example.

When the timeout is reached, the unit gets aborted and enters "failed" state. What happens then depends on the dependencies specified by "Wants" and "Requires". When the zvol target requires all zvol@.service and the zvol target is required by multi-user.target, then systemd will completely refuse to boot and go to recovery mode. This is *not* what we want but since zfs-volumes.target is only WantedBy=zfs.target, it only gets marked as failed.

c0d3z3r0 commented:
Currently it seems the generator does not create the target directory, but I don't know why, yet

aerusso commented:
@c0d3z3r0 Could you comment what you mean it "does not create the target directory"? Are you calling the generator manually? This was not something I ever experienced during testing.

c0d3z3r0 commented:
> @c0d3z3r0 Could you comment what you mean it "does not create the target directory"? Are you calling the generator manually? This was not something I ever experienced during testing.

Sorry, that was unclear - I meant the zvol_reqdir doesn't get created

aerusso commented:
@c0d3z3r0 I'm very confused how that's possible. Is the `local-fs.target.wants` being populated with the links to the mount files? If you're calling it manually, are you passing `/run/systemd/generator` as its first argument?

c0d3z3r0 commented:
> @c0d3z3r0 I'm very confused how that's possible. Is the `local-fs.target.wants` being populated with the links to the mount files? If you're calling it manually, are you passing `/run/systemd/generator` as its first argument?

not calling it manually; I will do a fresh zfs setup later and test this again

aerusso commented:
@c0d3z3r0 I've actually think I have a simpler way to do all of this. Let me get that up before wasting time testing an older patch.

c0d3z3r0 commented:
> @c0d3z3r0 I've actually think I have a simpler way to do all of this. Let me get that up before wasting time testing an older patch.

great, lmk when it's ready for testing :)

aerusso commented:
@c0d3z3r0 Take a look at this version; I think you will be pleased.  It drops the extra template unit entirely, opting just to `Require=dev-zvol-%i.device` directly (via drop-in file). Cursory testing on my end shows that JobTimeoutSec=30 should match the existing behavior.

I also was unable to reproduce any issues creating the dependency directory here.

c0d3z3r0 commented:
@aerusso hah, very nice solution! congrats :)) LGTM

pzakha commented:
It seems like we are now depending on `.device` units being automatically created by systemd for each zvol. I was curious to see how this works in practice.

I've manually created a new zvol `rpool/zvol1` and looked for `dev-zvol-rpool-zvol1.device`, but it wasn't there. I've tried adding `Requires=dev-zvol-rpool-zvol1.device` to the `zfs-volumes.target` unit file, and that resulted in `zfs-volumes.target` to be inactive on reboot with `systemd[1]: Dependency failed for ZFS volumes are ready.`

Is there something else required to have `.device` files created? I do see that `/dev/zvol/rpool/zvol1` exists, and I do see `.device` files for other devices such as `/dev/xvda --> dev-xvda.device`. Are we missing some udev rules?

aerusso commented:
@pzakha I'm running systemd 243 (Debian unstable). I was worried I was going to have to do something, but creating the zvol "just worked for me." 

In `99-systemd.rules`, I found

    SUBSYSTEM=="block", TAG+="systemd"

and, `man systemd.device` claims that:

> systemd will dynamically create device units for all kernel devices that are marked with the "systemd" udev tag (by default all block and network devices, and a few others).

So this is supposedly "default" systemd behavior---does your distribution's man page reflect that?

jgallag88 commented:
@pzakha Were you testing this on a Delphix Engine? We disabled .device units for zvols because they were causing us to hit the hardcoded systemd limit for number of units on a system: https://github.com/delphix/delphix-platform/pull/80

pzakha commented:
Okay, thanks for the clarifications, I was not aware we had it disabled explicitly in the product, so sorry for the confusion.

So given that udev seems to already provide per-zvol units, which I believe was the main selling point for the original PR, what other advantages do we get with this new approach?

`zfs-zvol-wait.service` already waits on zvols to be ready, and filters out zvols that are not expected to have links (there are several corner cases). From my understanding, once we ship this PR, it will be redundant with `zfs-zvol-wait.service` and one of them would have to be removed.

kpande commented:
i am not a user of systemd but as i understand it the original purpose of those per-zvol unit was to grant the administrator the ability to wait for just the zvol they need to be available before continuing boot rather than waiting for *all* zvol to become ready.

pzakha commented:
So in this case, the idea would be to manually edit the FSLIST to include only the zvols that the administrator cares about and *not* use `history_event-zfs-list-cacher.sh`, right?

aerusso commented:
@pzakha 

>So in this case, the idea would be to manually edit the FSLIST to include only the zvols that the administrator cares about and not use `history_event-zfs-list-cacher.sh`, right?

This is not my intention with this PR. The idea is to have a "proper" integration into systemd, that "does the right thing" automatically. I read through your changes to `zvol_wait`, and this is not yet at feature parity (I'll have to fix it before it gets merged). In particular, it doesn't check `redact_snaps`

The way I see it, if people want to fine-tune their systemd unit dependencies, they should depend on specific `dev-zvol-*.device` units.  If they can't be bothered to do that, then just depend on the blanket `zfs-volumes.target`. We need the `zfs-mount-generator` framework for other features it provides, so ideally we push `zfs-volumes.target` management into the mount generator, and drop the `zvol_wait` infrastructure once all the kinks in this have been worked out (much later on).

pzakha commented:
Okay, that makes more sense. Once we bring this to feature parity (exclude the right zvols), it would make sense to remove `zfs-zvol-wait.service`. We may also want to extend the timeout to more than 30 seconds at that point as in practice it should not be required anyway. 

Keeping the `zvol_wait` script might still be worthwhile for cases where a pool is imported after boot has completed and we want to wait for the zvols to be complete.

This will still have the disadvantages of not scaling well, but I think it should be fine for most folks, and for others they can keep using `zvol_wait`.

Rudd-O commented:
Can the code, instead of using polling alone, use a combination of inotify/fanotify and polling, or perhaps just inotify/fanotify alone, since it knows ahead of time which volumes are supposed to appear?

Does this work well with ZFS encryption of subvolumes?

behlendorf commented:
@aerusso if you still want to move forward with this tighter systemd integration, can you get this rebased and add the missing functionality.

aerusso commented:
@behlendorf I'm going to wait for #12138 to land, rather than force it to rebase on this patch (or lose the race and then rebase this a second time).

behlendorf commented:
@aerusso so now that #12138 has landed, can I talk you in to dusting off this PR so we can move forward with it?

aerusso commented:
So, I have a (local, untested) version that is rebased on the new C version of `zfs-mount-generator`, with feature parity to what's currently here.  However: it does NOT have feature parity with mainlined `zvol_wait`.  In particular, the following filter is going to be tricky:

```bash
list_zvols() {
	read -r default_volmode < /sys/module/zfs/parameters/zvol_volmode
	zfs list -t volume -H -o \
	    name,volmode,receive_resume_token,redact_snaps |
	    while IFS="	" read -r name volmode token redacted; do # IFS=\t here!

		# /dev links are not created for zvols with volmode = "none"
		# or for redacted zvols.
		[ "$volmode" = "none" ] && continue
		[ "$volmode" = "default" ] && [ "$default_volmode" = "3" ] &&
		    continue
		[ "$redacted" = "-" ] || continue

		# We also ignore partially received zvols if it is
		# not an incremental receive, as those won't even have a block
		# device minor node created yet.
		if [ "$token" != "-" ]; then

			# Incremental receives create an invisible clone that
			# is not automatically displayed by zfs list.
			if ! zfs list "$name/%recv" >/dev/null 2>&1; then
				continue
			fi
		fi
		echo "$name"
	done
}
```

The problem is that `/sys/module/zfs/parameters/zvol_volmode` won't be (guaranteed to be) available at systemd generator-time.  And I really don't think adding snapshot-receive events to add to the already overburdened history-cacher zedlet is a good idea (and I have a suspicion I'd miss at least one weird corner case).

I'd like to bounce the following proposal off people: refactor zvol_wait to be able to act on a single dataset (or on all datasets, by default, with no command line options).  Then, call that from a systemd unit.  Unfortunately, this gets rid of a lot of the beautiful simplicity we achieved in the last version (where everything was handled nicely by systemd).  We also can't use `zvol_wait` (AFAICT) to just filter the units: there is no systemd "`ConditionExecSuccess`".

behlendorf reviewed (approved):
@pzakha @c0d3z3r0 it'd be great if you could take another look at this.

My feeling is relying on the updated version of `zvol_wait` is still a step in the right direction and it'll be nice to have this functionality available.  Even if it adds back some of the complexity.

pzakha commented:
I took a quick glimpse and the new logic seems to make sense to me. I'll try to do a full review later this week.
